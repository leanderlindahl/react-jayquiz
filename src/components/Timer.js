import React, { Component } from 'react';
import PropTypes from 'prop-types';

class Timer extends Component {
  constructor(props) {
    super(props);
    const { startValueInSeconds } = this.props;
    this.state = {
      secondsRemaining: startValueInSeconds
    };
  }

  componentDidMount() {
    this.timerID = setInterval(() => this.tick(), 1000);
  }

  componentDidUpdate(prevProps) {
    const { addSecondsAmount, reset, startValueInSeconds } = this.props;
    const { secondsRemaining } = this.state;
    if (reset !== prevProps.reset) {
      setInterval(
        this.setState({
          secondsRemaining: startValueInSeconds
        }),
        10000
      );
    }
    if (addSecondsAmount !== prevProps.addSecondsAmount) {
      this.setState({
        secondsRemaining: secondsRemaining + addSecondsAmount
      });
    }
  }

  componentWillUnmount() {
    clearInterval(this.timerID);
  }

  tick() {
    const { secondsRemaining } = this.state;
    const { onComplete, running } = this.props;
    if (running) {
      this.setState({
        secondsRemaining: secondsRemaining - 1
      });
      if (secondsRemaining - 1 <= 0) {
        return onComplete();
      }
    }
  }

  render() {
    const { secondsRemaining } = this.state;
    return <div>{`Timer: ${secondsRemaining}`}</div>;
  }
}

Timer.propTypes = {
  addSecondsAmount: PropTypes.number,
  onComplete: PropTypes.func,
  reset: PropTypes.string,
  running: PropTypes.bool,
  startValueInSeconds: PropTypes.number
};

Timer.defaultProps = {
  addSecondsAmount: 0,
  onComplete: () => 'Timer reached zero and stopped.',
  reset: '0',
  running: true,
  startValueInSeconds: 15
};

export default Timer;
