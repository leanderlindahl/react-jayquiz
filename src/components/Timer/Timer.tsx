import React, { Component } from 'react';

export interface TimerProps {
  addSecondsAmount: number;
  onComplete(): void;
  reset: string;
  running: boolean;
  startValueInSeconds: number;
}

export interface TimerState {
  secondsRemaining: number;
}

interface Timer {
  timerID: any;
}

class Timer extends Component<TimerProps, TimerState> {
  static defaultProps = {
    addSecondsAmount: 0,
    onComplete: () => 'Timer reached zero and stopped.',
    reset: '0',
    running: true,
    startValueInSeconds: 15
  };

  state = {
    secondsRemaining: this.props.startValueInSeconds
  };

  public componentDidMount() {
    console.log('componentDidMount');
    this.timerID = setInterval(() => {
      this.tick();
    }, 1000);
  }

  public componentDidUpdate(prevProps: TimerProps, prevState: TimerState) {
    const { addSecondsAmount, reset, startValueInSeconds } = this.props;
    const { secondsRemaining } = this.state;

    if (reset !== prevProps.reset) {
      this.setState({ secondsRemaining: startValueInSeconds });
    }

    if (addSecondsAmount !== prevProps.addSecondsAmount) {
      this.setState({
        secondsRemaining: secondsRemaining + addSecondsAmount
      });
    }
  }

  public componentWillUnmount() {
    clearInterval(this.timerID);
  }

  public tick() {
    const { secondsRemaining } = this.state;
    const { onComplete, running } = this.props;
    if (running) {
      this.setState({
        secondsRemaining: secondsRemaining - 1
      });
      if (secondsRemaining - 1 <= 0) {
        return onComplete();
      }
    }
  }

  public render() {
    const { secondsRemaining } = this.state;
    return <div>{`Timer: ${secondsRemaining}`}</div>;
  }
}

export default Timer;
