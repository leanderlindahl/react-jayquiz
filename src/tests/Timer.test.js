import React from 'react';
import renderer from 'react-test-renderer';
import Timer from '../components/Timer';

test('Timer counts down from a given value', () => {
  const component = renderer.create(<Timer startValueInSeconds={5} running />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
