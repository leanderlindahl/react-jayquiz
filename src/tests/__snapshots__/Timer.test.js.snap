// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer counts down from a given value 1`] = `
<div>
  Timer: 5
</div>
`;
